// js-util/lib/local-storage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var store = {
  data: {}, // Used to fake when store not available (ie. on server or JSDOM).

  removeItem: function removeItem(key) {
    if (global.localStorage) {
      global.localStorage.removeItem(key);
    } else {
      delete this.data[key];
    }
  },

  setItem: function setItem(key, value) {
    if (global.localStorage) {
      global.localStorage.setItem(key, value);
    } else {
      this.data[key] = value;
    }
  },

  getItem: function getItem(key) {
    if (global.localStorage) {
      return global.localStorage.getItem(key);
    } else {
      return this.data[key];
    }
  }
};

var LocalStorage = (function () {
  function LocalStorage() {
    _classCallCheck(this, LocalStorage);
  }

  _createClass(LocalStorage, [{
    key: "keys",

    /**
     * Retrieves the set of keys within local-storage.
     */
    value: function keys() {
      var storage = global.localStorage ? global.localStorage : store.data;
      return _lodash2["default"].keys(storage);
    }

    /*
    * Gets or sets the value for the given key.
    *
    * @param key:         The unique identifier of the value (this is prefixed with the namespace).
    * @param value:       (optional). The value to set (pass null to remove).
    * @param options:
    *           default:  (optional). The default value to return if the session does not contain the value (ie. undefined).
    *
    * @return the read value.
    */
  }, {
    key: "prop",
    value: function prop(key, value) {
      var options = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

      if (_lodash2["default"].isNull(value)) {
        // REMOVE.
        store.removeItem(key);
      } else if (!_lodash2["default"].isUndefined(value)) {

        // WRITE.
        var type;
        if (_lodash2["default"].isString(value)) {
          type = "string";
        } else if (_lodash2["default"].isBoolean(value)) {
          type = "bool";
        } else if (_lodash2["default"].isNumber(value)) {
          type = "number";
        } else if (_lodash2["default"].isDate(value)) {
          type = "date";
        } else {
          type = "object";
        }

        var writeValue = { value: value, type: type };
        store.setItem(key, JSON.stringify(writeValue));
      } else {

        // READ ONLY.
        var json = store.getItem(key);
        if (json) {
          json = JSON.parse(json);
          switch (json.type) {
            case "null":
            case "bool":
            case "string":
              value = json.value;
              break;

            case "number":
              value = parseFloat(json.value);
              break;

            case "date":
              value = new Date(json.value);
              break;

            case "object":
              value = json.value;
              break;
          }
        } else {
          value = undefined;
        }

        if (_lodash2["default"].isUndefined(value)) {
          value = options["default"];
        }
      }

      // Finish up.
      return value;
    }
  }]);

  return LocalStorage;
})();

exports["default"] = new LocalStorage();
module.exports = exports["default"];
